### PR TITLE
[MI-3214] Add script to create posts in channels, DMs and GMs

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: config/config.sample.json [hunk 1]
```diff
@@ -35,6 +35,9 @@
         }
     ],
     "PostsConfiguration": {
-        "Count": 100
+        "Count": 100,
+        "MaxSentenceLength": 10,
+        "MaxWordsLength": 10,
```
Comment (raghavaggarwal2308) [inline config/config.sample.json]:
```suggestion
        "MaxWordLength": 10,
```

File: k6/createPosts.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import http from 'k6/http';
+
+const config = JSON.parse(open('../config/config.json'));
+const response = JSON.parse(open('../temp_store.json'));
+
+export const options = {
+    vus: config.PostsConfiguration.Count,
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js:7]: what is the meaning of vus here?

File: k6/createPosts.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    duration: config.PostsConfiguration.Duration,
+}
+
+function getRandomMessage(sentenceLength, wordsLength) {
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]:
```suggestion
function getRandomMessage(sentenceLength, wordLength) {
```

File: k6/createPosts.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	let message = '';
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+	let words = 0;
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]: indentation

File: k6/createPosts.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+const response = JSON.parse(open('../temp_store.json'));
+
+export const options = {
+    vus: config.PostsConfiguration.Count,
+    duration: config.PostsConfiguration.Duration,
+}
+
+function getRandomMessage(sentenceLength, wordsLength) {
+	let message = '';
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]: We should give a better name to it, so that it explains what is the use of this

File: k6/createPosts.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	let words = 0;
+    sentenceLength = Math.floor(Math.random() * sentenceLength) + 1;
```
Comment (raghavaggarwal2308) [inline k6/createPosts.js]: I think this is the number of words in a sentence, but it sounds like its the number of letters in the sentence. Please change

File: scripts/create_channel.go [hunk 7]
```diff
@@ -55,6 +55,17 @@ func CreateChannels(config *serializers.Config, logger *zap.Logger) error {
 			continue
 		}
 
+		for count := 0; count < len(newUserIDs); count++ {
+			if _, _, err := client.AddChannelMember(createdChannel.Id, newUserIDs[count]); err != nil {
+				logger.Error("unable to add users to the channel",
```
Comment (raghavaggarwal2308) [inline scripts/create_channel.go]:
```suggestion
				logger.Error("Unable to add users to the channel",
```

File: scripts/create_users.go [hunk 8]
```diff
@@ -23,10 +23,17 @@ func CreateUsers(config *serializers.Config, logger *zap.Logger) error {
 			continue
 		}
 
+		_, userResponse, err := client.Login(user.Username, user.Password)
+		if err != nil {
+			logger.Info("unable to login new user", zap.String("username", user.Username), zap.Error(err))
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]:
```suggestion
			logger.Info("Unable to login new user", zap.String("Username", user.Username), zap.Error(err))
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: k6/createPosts.js (hunk 4 above)
Comment (ayusht2810) [reply k6/createPosts.js]: it is correct in code

File: k6/createPosts.js (hunk 2 above)
Comment (ayusht2810) [reply k6/createPosts.js:7]: virtual users

File: scripts/create_users.go [hunk 9]
```diff
@@ -18,19 +18,18 @@ func CreateUsers(config *serializers.Config, logger *zap.Logger) error {
 			Password: user.Password,
 		})
 		if err != nil {
-			logger.Info("unable to create new user", zap.String("username", user.Username), zap.Error(err))
+			logger.Info("Unable to create new user", zap.String("username", user.Username), zap.Error(err))
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]:
```suggestion
			logger.Info("Unable to create new user", zap.String("Username", user.Username), zap.Error(err))
```

File: scripts/create_users.go [hunk 10]
```diff
@@ -18,19 +18,18 @@ func CreateUsers(config *serializers.Config, logger *zap.Logger) error {
...
 			continue
 		}
 
 		_, userResponse, err := client.Login(user.Username, user.Password)
 		if err != nil {
-			logger.Info("unable to login new user", zap.String("username", user.Username), zap.Error(err))
+			logger.Info("Unable to login new user", zap.String("username", user.Username), zap.Error(err))
```
Comment (raghavaggarwal2308) [inline scripts/create_users.go]: same

File: config/config.sample.json [hunk 11]
```diff
@@ -37,7 +37,7 @@
     "PostsConfiguration": {
         "Count": 100,
         "MaxSentenceLength": 10,
```
Comment (raghavaggarwal2308) [inline config/config.sample.json]:
```suggestion
        "MaxWordsCount": 10,
```

**Review by raghavaggarwal2308 — APPROVED** [review]

File: k6/createPosts.js (hunk 2 above)
Comment (raghavaggarwal2308) [reply k6/createPosts.js:7]: I think we should change the name. Ex. VirtualUsersCount or usersCount

Comment (ayusht2810) [reply k6/createPosts.js:7]: discussed

File: k6/createPosts.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	let message = '';
+    const characterSet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+	let words = 0;
+    wordsCount = Math.floor(Math.random() * wordsCount) + 1;
+    while (words < wordsCount) {
+        let count = 0;
+        wordLength = Math.floor(Math.random() * wordLength) + 1;
+        while (count < wordLength) {
+            message += characterSet.charAt(Math.floor(Math.random() * characterSet.length));
+            count += 1;
```
Comment (manojmalik20) [inline k6/createPosts.js]:
```suggestion
            count++;
```
Use this instead of `+= 1` everywhere.

File: k6/createPosts.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import http from 'k6/http';
+
+const config = JSON.parse(open('../config/config.json'));
+const response = JSON.parse(open('../temp_store.json'));
```
Comment (manojmalik20) [inline k6/createPosts.js]: Don't use variable name `response`.

File: k6/createPosts.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+        message += ' ';
+        words += 1;
+    }
+
+	return message;
+}
+
+function getRandomToken() {
+    let token = [];
```
Comment (manojmalik20) [inline k6/createPosts.js]:
```suggestion
    let tokens = [];
```

File: k6/createPosts.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    response.UserResponse.map((u) => token.push(u.token));
+    return token[(Math.floor(Math.random() * token.length))];
+}
+
+function getRandomChannel() {
+    let channelId = [];
```
Comment (manojmalik20) [inline k6/createPosts.js]:
```suggestion
    let channelIds = [];
```

File: scripts/create_channel.go [hunk 16]
```diff
@@ -47,13 +47,24 @@ func CreateChannels(config *serializers.Config, logger *zap.Logger) error {
 		}
 
 		if _, _, err := client.AddTeamMembers(team.Id, newUserIDs); err != nil {
-			logger.Error("unable to add users to the team", zap.String("TeamName", channel.MMTeamName), zap.Error(err))
+			logger.Error("Unable to add users to the team", zap.String("TeamName", channel.MMTeamName), zap.Error(err))
 			continue
 		}
 
+		for count := 0; count < len(newUserIDs); count++ {
```
Comment (manojmalik20) [inline scripts/create_channel.go]: Use range loop.

File: scripts/create_dm_and_gm.go [hunk 17]
```diff
@@ -10,13 +12,17 @@ import (
...
-	if _, _, err := client.Login(config.ConnectionConfiguration.AdminEmail, config.ConnectionConfiguration.AdminPassword); err != nil {
+	response, err := utils.LoadCreds()
+	if err != nil {
 		return err
 	}
 
-	response, err := utils.LoadCreds()
-	if err != nil {
+	if len(response.UserResponse) == 0 {
+		return errors.New("no new user created")
```
Comment (manojmalik20) [inline scripts/create_dm_and_gm.go:21]: We should use `GetOrCreateUser` for users.

File: scripts/create_dm_and_gm.go [hunk 18]
```diff
@@ -10,13 +12,17 @@ import (
...
+	}
+
+	client := model.NewAPIv4Client(config.ConnectionConfiguration.ServerURL)
+	if _, _, err := client.Login(config.UsersConfiguration[0].Username, config.UsersConfiguration[0].Password); err != nil {
```
Comment (manojmalik20) [inline scripts/create_dm_and_gm.go]: Remove username and use email. Also, what's the need for login here?

File: scripts/create_users.go [hunk 19]
```diff
@@ -47,3 +49,25 @@ func CreateUsers(config *serializers.Config, logger *zap.Logger) error {
 
 	return nil
 }
+
+func GetOrCreateUser(client *model.Client4, userDetails serializers.UsersConfiguration) (*model.User, error) {
```
Comment (manojmalik20) [inline scripts/create_users.go]:
```suggestion
func GetOrCreateUser(client *model.Client4, userDetails *serializers.UsersConfiguration) (*model.User, error) {
```
Use pointers everywhere.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**